### PR TITLE
Keysight Technologies 34980A Switch Matrix library

### PR DESCRIPTION
**Pull request by Neel-Sancheti**
Keysight Technologies 34980A Multifunction Switch/Measure Unit with support for Keysight technologies 34923A Multifunction Switch/Measure Unit.

### CODE REVIEW

**Comment by Neel-Sancheti** [comment]
As per Christoph Buchner's comment. I have added a commit and am waiting for the approval. I am new to git, all the guidance is appreciated.

Thankyou!

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:52]: See if you can convert the methods into `control`, `measurement`, or `setting` properties.

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:54]: This documentation string can better go into a docstring of one of the properties, or if a method cannot be converted into a property into the docstring format of that method, e.g.:
```
def some_method(self):
    """ Put the docstring over here
    Can also be multiline
    """

    # body of the method
```

This way, the documentation system that generates all the documentation knows where to find these docstrings and can use it for the doc-page of the instrument.

**Review by CasperSchippers — CHANGES_REQUESTED** [review]
Thanks for the contribution.
The PyMeasure instruments section provides a number of features that make programming and maintaining an instrument easier. I've remarked upon a few in the file, but do also reccomend you to take a look at this part of the documentation: https://pymeasure.readthedocs.io/en/latest/dev/adding_instruments.html.

A lot of properties can converted into a `control`, `measurement`, or `setting` property. For those that cannot be converted into this, I would recommend using `self.write` in stead of `self.switch_data.write` and `self.values` (or in some cases `self.ask`) in stead of `self.switch_data.query` (though, if at all possible, it is always recomended to use the aforementioned properties).

Apart from that, you could also have a look at how the Keithley 2700 or 2750 is implemented, these are also matrix multiplexers and might be helpful to look at.

Finally, don't forget to add the instrument to __init__.py file in the Keysight folder, add the required documentation files in the docs/api/instruments/keysight folder, and add yourself to the author lists (one at AUTHORS.txt at the root of the repo and one in docs/about/authors.rst).

If you have any questions or anything is not (completely) clear, feel free to let me know!

**Comment by Neel-Sancheti** [comment]
Hey Casper, I have updated the code according to your comments. Thank you for your help!

**Comment by CasperSchippers** [comment]
> Hey Casper, I have updated the code according to your comments. Thank you for your help!

That's very good; you might want to push the changes, than I can have another look.

**Comment by Neel-Sancheti** [comment]
> > Hey Casper, I have updated the code according to your comments. Thank you for your help!
> 
> That's very good; you might want to push the changes, than I can have another look.

I have updated them. Can you see the changes?

**Comment by CasperSchippers** [comment]
> I have updated them. Can you see the changes?

Yes, I'll have a look asap

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:463]: I think these (or at least most of these) can be nicely converted into a properties (also holds for a number of other methods, but the ones over here seem like a good starting point): for example `Instrument.measurement` for the measure_xx methods and `Instrument.control` for set_temperature_unit and is_temperature_unit.

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:562]: I see you often have such a construction; could you explain why this is needed? Is setting a value once not sufficient?

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:358]: The arguments (chan, resolution and range_param) can be removed, right? they're not used anyway

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:387]: I would suggest setting the default value to `None` rather than `""` and change this line into `chan is None`; I think it's a bit neater

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:384]: for methods with arguments it is always nice to have at the end of the docstring a list of the parameters in the following format:
```
:param chan: A optional string parameter that defines the channel for which the temperature unit is set.
:param unit: An optional string parameter that defines which unit to set.
```

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:405]: Structures like this (i.e. writing a certain query, reading the response and casting it into a certain type) can be best performed using a `self.ask` (in case you expect a single string to be returned) or `self.values` (in case you expect one or more numeric values to be returned).

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (CasperSchippers) [inline pymeasure/instruments/keysight/keysight34980A.py:409]: The fact that the instrument has an internal DMM or a channel for a lot of commands complicates the implementation quite a bit; I'm wondering if there would be a neater way to implement this.

You could consider defining an `Instrument.control` property for the temperature unit of the DMM and have a sub-class for the different channels such that you can call Keysight34980A34980A.ch1.temperature_unit (see for example the KeysightDSOX1102G class where there are 2 channels).

If there are really many channels, you could also consider another construction where you have a sort of channelmanager that manages the channels such that in the end you can call it like Keysight34980A34980A.ch(1).temperature_unit:
```
class Keysight34980A34980A:
    def __init__(self):
        self.ch = ChannelManager()


class ChannelManager:
    def __init__(self):
        self.channel = 0

    def __call__(self, channel):
        self.channel = channel
        return self

    def do_something(self):
        print(f"doing something to channel {self.channel}")


a = Keysight34980A34980A()
a.ch(1).do_something()
a.ch(2).do_something()

# Calling without the number will use the last entered channel number
a.ch.do_something()

# You can't just store a channel, because you constantly refer to the same object
ch1 = a.ch(1)
ch2 = a.ch(2)
ch1.do_something()
# but you can using copy if you'd want
import copy
ch1 = copy.copy(a.ch(1))
ch2 = copy.copy(a.ch(2))
ch1.do_something()
ch2.do_something()
```

File: pymeasure/instruments/keysight/keysight34980A.py
Comment (Neel-Sancheti) [reply pymeasure/instruments/keysight/keysight34980A.py:562]: It is just a verification to check that the value is set correctly.

Comment (CasperSchippers) [reply pymeasure/instruments/keysight/keysight34980A.py:562]: Do you know if it happens often/at all that the value isn't correctly set? I'd think that it would not be required in general and that it just clutters the communication. Moreover, if there is another reason a certain value cannot be set, it will try again indefinitely.

**Comment by AidenDawn** [comment]
From what I know of this instrument, this pull request probably needs quite a lot of restructuring before it could be merged. The mainframe contains a DMM (or not, I think thats an option) and the control hardware for lots of different submodule types which are mostly relay matrices but also include some digital IO and on the more recent hardware revisions can contain digitizers too. Happy to have a look into this a bit more at some point but I believe storing the channel/module functions inside the main mainframe class would be oversimplifying the complexity of this instrument

**Comment by bmoneke** [comment]
Hey @Neel-Sancheti do you intend to finish this contribution (thanks for opening it)?
If not, we can close it. The code will stay in case you want to continue the work later on.

**Comment by Neel-Sancheti** [comment]
Hello,

Sorry for the delayed response. For now we can close it. I will come back
to it later. Or someone else can continue the work.

Thankyou!

**Comment by bmoneke** [comment]
Thanks for your response.

Closed until someone wants to pick the work up again.

**Pull request closed without merging**
